### PR TITLE
Implement reversion of nodes after editing via searcher.

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description
Add functionality to revert a node that was changed during editing through the searcher. 
Note that currently, nodes are not edited by the searcher, except upon confirmation, Thus this functionality results in a no-op at the moment.

### Important Notes
[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

### CODE REVIEW

**Review by farmaazon — COMMENTED** [review]
Actually, as I've checked the aborting editing... do not revert the expression at the moment. I think this is a regression. So we actually _can_ check if this branch fixes it.

**Comment by MichaelMauderer** [comment]
> Actually, as I've checked the aborting editing... do not revert the expression at the moment. I think this is a regression. So we actually _can_ check if this branch fixes it.

After checking, it appears that this is not fixed by this PR at the moment. Only the source code is reverted, the view logic was not changed and seems not to revert right away, only after some other update triggers a re-evaluation of the view expression.

**Review by farmaazon — APPROVED** [review]

**Comment by MichaelMauderer** [comment]
> > Actually, as I've checked the aborting editing... do not revert the expression at the moment. I think this is a regression. So we actually _can_ check if this branch fixes it.
> 
> After checking, it appears that this is not fixed by this PR at the moment. Only the source code is reverted, the view logic was not changed and seems not to revert right away, only after some other update triggers a re-evaluation of the view expression.

The tracking issue is filed here: https://www.pivotaltracker.com/story/show/182926584
This should be fixed in a later stage of the user story that reworks some of the related functionality.

<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/182926584">Aborting editing via `Esc` does not correctly update node view - Pivotal Tracker</a></strong></div></blockquote>

File: app/gui/src/controller/searcher.rs [hunk 1]
```diff
@@ -1235,16 +1244,63 @@ impl EditMetadataGuard {
...
+            }),
+        )?;
+        Ok(edit_status)
+    }
+
+    fn revert_node_expression(&self) -> FallibleResult {
+        let edit_status = self.get_saved_expression()?;
+        match edit_status {
+            None => {
+                tracing::warn!("Tried to revert edited node, but found not edit metadata.");
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: Please check the sentences with gramarly or something similar. This should be "... , but no edit metadata was found.". Also, what does it mean "revert edited node". You are not reverting the node, rather reverting its expression to the previous state, right? For example, node icons are not reverted? I believe that naming / explanations should be improved here and possibly in other places too.

File: app/gui/src/controller/searcher.rs [hunk 2]
```diff
@@ -1235,16 +1244,63 @@ impl EditMetadataGuard {
...
+            }
+            Some(NodeEditStatus::Created) => {
+                tracing::debug!("Deleting node {} after edit reverted.", self.node_id);
+                self.graph.graph().remove_node(self.node_id)?;
+            }
+            Some(NodeEditStatus::Edited { previous_expression }) => {
+                tracing::debug!(
+                    "Reverting node {} after edit reverted. Expression to revert: {}",
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: What does mean "Reverting node {} after edit reverted"? I do not understand this sentence TBH. Also, is the "Expression to revert" the old one or the new one?

File: app/gui/src/controller/searcher.rs [hunk 3]
```diff
@@ -1235,16 +1244,63 @@ impl EditMetadataGuard {
             }),
         )
     }
+
+    fn get_saved_expression(&self) -> FallibleResult<Option<NodeEditStatus>> {
+        let module = &self.graph.graph().module;
+        let mut edit_status = None;
+        module.with_node_metadata(
+            self.node_id,
+            Box::new(|m| {
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: `m` is really bad variable name. Does it refer to "module"? (this is my first guess).
Edit: Ahh, it is "metadata" – IMO if the reference to the expanded name is not in the same line or in the previous line, the lambda should use more expresive names.

File: app/gui/src/controller/searcher.rs [hunk 4]
```diff
@@ -1235,16 +1244,63 @@ impl EditMetadataGuard {
...
-            tracing::error!(
+        if self.revert_expression.get() {
+            self.revert_node_expression().unwrap_or_else(|e| {
+                tracing::error!(
+                    "Failed to revert node edit after editing ended because of an error: {}",
+                    e
+                )
+            });
+        } else {
+            tracing::debug!("Not reverting node expression after edit.")
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs:1297]: Here the sentence is correct! It tells about reverting EXPRESSION, not the node.

File: app/gui/src/controller/searcher.rs [hunk 5]
```diff
@@ -2278,16 +2334,21 @@ pub mod test {
...
+        let node = graph.nodes().unwrap().last().unwrap().clone();
+        let initial_node_expression = node.main_line.expression().clone();
         let node_id = node.info.id();
         searcher.mode = Immutable(Mode::EditNode { node_id });
-        searcher.node_metadata_guard =
-            Rc::new(Some(EditMetadataGuard::new(node_id, searcher.graph.clone_ref())));
+        searcher.node_edit_guard =
+            Rc::new(Some(EditGuard::new(node_id, searcher.graph.clone_ref())));
+
+        // Edit node
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: This is not a sentence.

File: app/gui/src/controller/searcher.rs [hunk 6]
```diff
@@ -2312,5 +2373,32 @@ pub mod test {
...
+    fn edit_guard_no_revert() {
+        let Fixture { test: _test, mut searcher, .. } = Fixture::new();
+        let graph = searcher.graph.graph();
+        let node = graph.nodes().unwrap().last().unwrap().clone();
+        let node_id = node.info.id();
+        searcher.mode = Immutable(Mode::EditNode { node_id });
+        searcher.node_edit_guard =
+            Rc::new(Some(EditGuard::new(node_id, searcher.graph.clone_ref())));
+
+        // Edit node
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: not a sentence

**Review by wdanilo — CHANGES_REQUESTED** [review]

**Review by wdanilo — APPROVED** [review]

**Pull request merged**